src/osd/osd_types.h

```diff
@@ -2690,31 +2692,66 @@ inline ostream& operator<<(ostream& out, const pg_log_t& log)
  *  also used to pass missing info in messages.
  */
 struct pg_missing_t {
+  struct old_item {
+    eversion_t need, have;
+    old_item() {}
+    old_item(eversion_t n, eversion_t h) : need(n), have(h) {}
+
+    void encode(bufferlist& bl) const {
+      ::encode(need, bl);
+      ::encode(have, bl);
+    }
+    void decode(bufferlist::iterator& bl) {
+      ::decode(need, bl);
+      ::decode(have, bl);
+    }
+  };
+  WRITE_CLASS_ENCODER(old_item)
+
   struct item {
     eversion_t need, have;
-    item() {}
-    item(eversion_t n) : need(n) {}  // have no old version
-    item(eversion_t n, eversion_t h) : need(n), have(h) {}
+    bool can_recover_partial;
+    interval_set<uint64_t> dirty_extents;
+    item() : can_recover_partial(false) {}
+    item(eversion_t n, eversion_t h) : need(n), have(h), can_recover_partial(false) {}
+    item(const pg_log_entry_t& e) : need(e.version), have(e.prior_version),
+	 can_recover_partial(e.can_recover_partial) {
+      if (can_recover_partial)
+        dirty_extents.insert(e.dirty_extents);
+    }
 
+    void update(const pg_log_entry_t& e) {
+      need = e.version;
+      can_recover_partial = (can_recover_partial && e.can_recover_partial);
+      dirty_extents.union_of(e.dirty_extents);
+    }
     void encode(bufferlist& bl) const {
       ::encode(need, bl);
```

Comment:
@athanatos ,  encoding process of struct item is wrapped by the encoding process of struct pg_missing_t. And pg_missing_t uses ENCODE_START, ENCODE_FINISH macros to deal with the comparability.